clients/venice-producer/src/main/java/com/linkedin/venice/producer/AbstractVeniceProducer.java

```diff
@@ -0,0 +1,445 @@
+      Properties veniceWriterProperties) {
+    int amplificationFactor = versionCreationResponse.getAmplificationFactor();
+    Integer partitionCount = versionCreationResponse.getPartitions();
+    Properties partitionerProperties = new Properties();
+    partitionerProperties.putAll(versionCreationResponse.getPartitionerParams());
+    VenicePartitioner venicePartitioner = PartitionUtils.getVenicePartitioner(
+        versionCreationResponse.getPartitionerClass(),
+        amplificationFactor,
+        new VeniceProperties(partitionerProperties));
+    return constructVeniceWriter(
+        veniceWriterProperties,
+        new VeniceWriterOptions.Builder(versionCreationResponse.getKafkaTopic()).setPartitioner(venicePartitioner)
+            .setPartitionCount(partitionCount)
+            .setChunkingEnabled(false)
+            .build());
+  }
+
+  // Visible for testing
+  protected VeniceWriter<byte[], byte[], byte[]> constructVeniceWriter(
+      Properties properties,
+      VeniceWriterOptions writerOptions) {
+    return new VeniceWriterFactory(properties).createVeniceWriter(writerOptions);
+  }
+
+  protected RecordSerializer<Object> getSerializer(Schema schema) {
+    return FastSerializerDeserializerFactory.getAvroGenericSerializer(schema);
+  }
+
+  private static Schema getSchemaFromObject(Object object) {
+    if (object instanceof GenericContainer) {
+      GenericContainer obj = (GenericContainer) object;
+      return obj.getSchema();
+    } else if (object instanceof CharSequence) { // convenience option.
+      return STRING_SCHEMA;
+    } else if (object instanceof Integer) {
+      return INT_SCHEMA;
+    } else if (object instanceof Long) {
+      return LONG_SCHEMA;
+    } else if (object instanceof Double) {
+      return DOUBLE_SCHEMA;
+    } else if (object instanceof Float) {
+      return FLOAT_SCHEMA;
+    } else if (object instanceof byte[] || object instanceof ByteBuffer) {
+      return BYTES_SCHEMA;
+    } else if (object instanceof Boolean) {
+      return BOOL_SCHEMA;
+    } else {
+      throw new VeniceException(
+          "Venice Producer only supports Avro objects, and primitives, found object of class: "
+              + object.getClass().toString());
+    }
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(K key, V value) {
+    String error = validateProducer();
+    if (!StringUtils.isEmpty(error)) {
+      return getFutureCompletedExceptionally(error);
+    }
+    return asyncPutInternal(APP_DEFAULT_LOGICAL_TS, key, value);
+  }
+
+  @Override
+  public CompletableFuture<DurableWrite> asyncPut(long logicalTime, K key, V value) {
+    String error = validateProducerWithLogicalTime(logicalTime);
+    if (!StringUtils.isEmpty(error)) {
+      return getFutureCompletedExceptionally(error);
+    }
+    return asyncPutInternal(logicalTime, key, value);
+  }
+
+  private CompletableFuture<DurableWrite> asyncPutInternal(long logicalTime, K key, V value) {
```

Comment:
So IIUC, the difference between `asyncPut(K, V)` and `asyncPut(time, K, V)` is just that `validateProducerWithLogicalTime` does an extra check compared to `validateProducer`. I don't think it's worth tying ourselves into knots over this extra check. We could make the code simpler IMHO if `asyncPut(time, K, V)` contains the full check as well as the full content of `asyncPutInternal`, and `asyncPut(K, V)` just delegates to `asyncPut(APP_DEFAULT_LOGICAL_TS, K, V)`. We would do an extra boolean check on the time-free API as compared to the currently proposed implementation, but I don't think that's a big deal in this context. The online apps using this shouldn't be churning out millions of writes per sec per node, so these minor details probably don't matter here.

If you agree, then the same refactoring could apply to the update and delete functions as well.

Reply:
Yeah, that sounds okay. I did it this way to order the error messages to check producer state then validate the logical_ts. But I don't mind simplifying code and change the order of checks

Reply:
The issue is that `asyncPut(K, V)` calls `asyncPut(APP_DEFAULT_LOGICAL_TS, K, V)`. Since `APP_DEFAULT_LOGICAL_TS` is = -2, it will end up failing the if check. So, I decided to add the `asyncPutInternal` without any checks.

I can move the `validateProducer` call inside of the `asyncPutInternal` function and leave the `logicalTs` validation in the APIs with `logicalTs`. Does this work?

Reply:
```java
@Override
public CompletableFuture<DurableWrite> asyncPut(K key, V value) {
  return asyncPutInternal(APP_DEFAULT_LOGICAL_TS, key, value);
}

@Override
public CompletableFuture<DurableWrite> asyncPut(long logicalTime, K key, V value) {
  if (logicalTime < 0) {
    return getFutureCompletedExceptionally("Logical time must be a non-negative value. Got: " + logicalTime);
  }

  return asyncPutInternal(logicalTime, key, value);
}

private CompletableFuture<DurableWrite> asyncPutInternal(long logicalTime, K key, V value) {
  String error = validateProducer();
  if (!StringUtils.isEmpty(error)) {
    return getFutureCompletedExceptionally(error);
  }
  ...
}

```